docs/proposals/trimming-friendly-registrar-proposal.md

```diff
@@ -0,0 +1,483 @@
+
+### Option B: Pre-trimming + post-trimming MSBuild tasks
+
+We would modify the DLLs directly before the code is AOTed. Instead of ILLink and custom linker steps, we would create
+an independent MSBuild task to perform the weaving. This would include generating the `__Registrar_Callback__` nested
+classes with UCO endpoints, proxy classes for generic types, and generating any `[DynamicDependency]` or other attributes
+to express all inverse dependencies. At the same time, we would generate the Objective-C code and any information about
+native dependencies (frameworks, static libraries, other generated Objective-C source files).
+
+It should only be necessary to pre-process any third-party libraries and the runtimepack DLLs once after restore and not
+with every publish.
+
+The post-trimming task would analyze the object file produced by ILC and collect all the Objective-C source files
+based on the registrar callbacks that survived trimming and their dependencies. These source files will be compiled
+and linked with the AOTed binary.
+
+- **Pros**:
+  - generated code isn't distributed with runtimepack/nupkg
+    - it would be an invisible change for libraries authors - they won't need to rebuild their apps for .NET 9
+  - when we generate IL directly, we have control over visibility of certain members and nested types if needed
+  - we can reuse some code from the current codebase if we implement the pre-trimming task using Mono.Cecil
+- **Cons**:
+  - slower than source generators - all code is generated when publishing
+
+### Option C: Source generator + post-trimming MSBuild task
+
+The code for the UCO callbacks should be expressible in C#. The only problem that we might have to work around is the
+absence of the "inverse dependency" attributes. It might be a good use case to drive the implementation of this feature.
+
+The post-trimming task would be exactly the same as in Option B.
+
+- **Pros**:
+  - likely faster than an IL-weaving MSBuild task
+    - code is generated while editing the source files
+    - the code for the xamarin runtime and for all libraries (e.g, MAUI) would already be pre-built in managed DLLs
+  - code generator can produce errors and warnings while editing the source files for better DX
+  - source generators are more future-proof than Mono.Cecil
+- **Cons**:
+  - the generated code has to be distributed with runtimepack/nupkg
+    - all libraries authors will have to publish their code specifically for .NET 9 and any future releases
+      which change the registrar
+  - generating Objective-C code in roslyn source generators is a _hack_ that isn't officially supported
+    - source generators are _synchronous_ and doing many IO operations will slow them down, causing lagging in the editor
+  - its's not yet clear how to correctly generate the `[DynamicDependency]` attributes and if the inverse dependency
+    attributes could be added to .NET
+
```

Comment:
I would also be worried about the cost of using pinvoke/reverse pinvoke. Have you measured the difference compared to the existing solution?